add a few navigation shortcuts

robtfm opened the pull request:
adds support for (configurable) word left/right and line start/end navigation.

i'm in the process of removing egui from my app, and text entry is the main blocker. to that end i hope to also add support for
- copy/paste
- multiline entry
- selection hilighting
- mouse interactions (though i suspect this won't be upstreamed easily)

i fully understand if you think this pr and these other features are out of scope of a "simple" text entry field though.

rparrett commented on src/lib.rs:
```diff
@@ -187,6 +188,91 @@ pub struct TextInputSettings {
+#[cfg(target_os = "macos")]
+impl Default for TextInputNavigationBindings {
+    fn default() -> Self {
+        fn default() -> Self {
```
CI isn't testing on macOS at the moment, but it's failing here.

robtfm replied:
apols, messed up the merge

rparrett commented:
Cool. At a glance I feel like this fits in well. I'll take a closer look soon.

> i fully understand if you think this pr and these other features are out of scope of a "simple" text entry field though.

Let me give that a bit amount of thought. The evolving "bevy development roadmap landscape" may have changed my thoughts a bit. `cosmic-text` may make some of those features less intense. But I still think there's a hard line at dependencies on crates that depend on Bevy.

rparrett commented:
As implemented here, there's an issue where e.g. `KeyCode::KeyA` and `KeyCode::Home` arrive on the same frame, the `KeyCode::KeyA` event is not processed.

But more generally, it's not possible to know which key arrived first unless we deal with a single stream of events. This sort of bug was my bane from Bevy versions 0.4 -> 0.13 when it finally became possible to fix it properly. So I would strongly prefer to do our processing on a single event stream. I think that a way around this might be to roughly:

- Move event processing into a separate system scheduled before the current `keyboard` system.
- Restructure `TextInputNavigationBindings` such that the key and modifier are stored separately
- In the event processing code, I think it *might* be okay to rely on `ButtonInput` purely for the modifiers. But I haven't totally thought that through and tracking key press state probably wouldn't be too much trouble if we have to.
- Make `TextInputAction` an `Event` and add `CharLeft` and `CharRight` and emit it
- Rework the current `keyboard` system so that it reacts to those events

I am open to other ideas but I think that would even simplify things overall. We wouldn't need the manual event reader that was added in https://github.com/rparrett/bevy_simple_text_input/pull/58, for instance.

robtfm commented:
I guess having a primary button for the action would avoid moving twice if you hold right and press the modifier twice too. Ok let me try.

robtfm commented:
i added modifiers and moved the char left/right and backspace+delete into the enum. 

i can make a separate system if you like, but i'm not sure the benefit is that much (and cloning the reader is super-cheap, it's just a usize)

rparrett commented:
Looks great, thanks!
